File: src/main/java/dev/esophose/discordbot/command/commands/misc/MoveCommand.java
package dev.esophose.discordbot.command.commands.misc;

import dev.esophose.discordbot.Sparky;
import dev.esophose.discordbot.command.DiscordCommand;
import dev.esophose.discordbot.command.DiscordCommandMessage;
import dev.esophose.discordbot.manager.CommandManager;
import dev.esophose.discordbot.utils.BotUtils;
import dev.esophose.discordbot.webhook.WebhookUtils;
import discord4j.core.object.Embed;
import discord4j.core.object.entity.Member;
import discord4j.core.object.entity.channel.GuildChannel;
import discord4j.core.object.entity.channel.GuildMessageChannel;
import discord4j.core.object.util.Permission;
import discord4j.core.object.util.PermissionSet;
import discord4j.core.object.util.Snowflake;
import java.util.Collections;
import java.util.List;
import reactor.core.publisher.Mono;

public class MoveCommand extends DiscordCommand {

    public MoveCommand(Sparky bot) {
        super(bot);
    }

    public void execute(DiscordCommandMessage message, Snowflake messageId, GuildChannel textChannel) {
        CommandManager commandManager = this.bot.getManager(CommandManager.class);

        if (!(textChannel instanceof GuildMessageChannel)) {
            commandManager.sendResponse(message.getChannel(), "Invalid text channel", "The target channel must be a text channel.").subscribe();
            return;
        }

        GuildMessageChannel targetChannel = (GuildMessageChannel) textChannel;
        message.getChannel().flatMap(channel -> channel.getMessageById(messageId).doOnError(error -> commandManager.sendResponse(message.getChannel(), "Failed to move message", "The given message snowflake was invalid.").subscribe())).subscribe(targetMessage -> {
            Mono.zip(targetMessage.getAuthorAsMember(), targetMessage.getAuthorAsMember().flatMap(Member::getAvatar))
                    .flatMap(specDetails -> WebhookUtils.createAndExecuteWebhook(targetChannel, specDetails.getT1().getDisplayName(), specDetails.getT2(), spec -> {
                        targetMessage.getContent().ifPresent(spec::setContent);
                        targetMessage.getAttachments().forEach(file -> spec.addFile(file.getFilename(), BotUtils.getAttachment(file.getUrl())));
                        if (!targetMessage.getEmbeds().isEmpty()) {
                            Embed embed = targetMessage.getEmbeds().get(0);
                            spec.setEmbed(embedSpec -> {
                                embed.getAuthor().ifPresent(author -> embedSpec.setAuthor(author.getName(), author.getUrl(), author.getIconUrl()));
                                embed.getColor().ifPresent(embedSpec::setColor);
                                embed.getDescription().ifPresent(embedSpec::setDescription);
                                embed.getFields().forEach(field -> embedSpec.addField(field.getName(), field.getValue(), field.isInline()));
                                embed.getFooter().ifPresent(footer -> embedSpec.setFooter(footer.getText(), footer.getIconUrl()));
                                embed.getImage().map(Embed.Image::getUrl).ifPresent(embedSpec::setUrl);
                                embed.getThumbnail().map(Embed.Thumbnail::getUrl).ifPresent(embedSpec::setThumbnail);
                                embed.getTimestamp().ifPresent(embedSpec::setTimestamp);
                                embed.getTitle().ifPresent(embedSpec::setTitle);
                                embed.getUrl().ifPresent(embedSpec::setUrl);
                            });
                        }
                    })).doOnSuccess(v -> {
                targetMessage.delete("Moved message").subscribe();
                commandManager.sendResponse(message.getChannel(), "Message moved", "The message was moved successfully.").subscribe();
            }).subscribe();
        });
    }

    @Override
    public String getName() {
        return "move";
    }

    @Override
    public List<String> getAliases() {
        return Collections.emptyList();
    }

    @Override
    public String getDescription() {
        return "Moves a message from one channel to another";
    }

    @Override
    public PermissionSet getRequiredBotPermissions() {
        return PermissionSet.of(Permission.SEND_MESSAGES, Permission.MANAGE_CHANNELS, Permission.MANAGE_MESSAGES, Permission.MANAGE_WEBHOOKS);
    }

    @Override
    public Permission getDefaultRequiredMemberPermission() {
        return Permission.MANAGE_MESSAGES;
    }

}

File: src/main/kotlin/dev/esophose/discordbot/command/commands/misc/MoveCommand.kt
package dev.esophose.discordbot.command.commands.misc

import dev.esophose.discordbot.Sparky
import dev.esophose.discordbot.command.DiscordCommand
import dev.esophose.discordbot.command.DiscordCommandMessage
import dev.esophose.discordbot.manager.CommandManager
import dev.esophose.discordbot.utils.BotUtils
import discord4j.common.util.Snowflake
import discord4j.core.`object`.entity.channel.GuildChannel
import discord4j.core.`object`.entity.channel.GuildMessageChannel
import discord4j.rest.util.Permission
import discord4j.rest.util.PermissionSet
import reactor.core.publisher.Mono

class MoveCommand : DiscordCommand() {

    override val name: String
        get() = "move"

    override val aliases: List<String>
        get() = emptyList()

    override val description: String
        get() = "Moves a message from one channel to another"

    override val requiredBotPermissions: PermissionSet
        get() = PermissionSet.of(Permission.SEND_MESSAGES, Permission.MANAGE_CHANNELS, Permission.MANAGE_MESSAGES, Permission.MANAGE_WEBHOOKS)

    override val defaultRequiredMemberPermission: Permission
        get() = Permission.MANAGE_MESSAGES

    fun execute(message: DiscordCommandMessage, messageId: Snowflake, textChannel: GuildChannel) {
        val commandManager = Sparky.getManager(CommandManager::class)

        if (textChannel !is GuildMessageChannel) {
            commandManager.sendResponse(message.channel, "Invalid text channel", "The target channel must be a text channel.").subscribe()
            return
        }

        message.channel.flatMap { channel -> channel.getMessageById(messageId).doOnError { commandManager.sendResponse(message.channel, "Failed to move message", "The given message snowflake was invalid.").subscribe() } }.subscribe { targetMessage ->
            Mono.zip(targetMessage.authorAsMember, targetMessage.authorAsMember.flatMap { it.avatar })
                    .flatMap { specDetails ->
                        textChannel.createWebhook { spec ->
                            spec.setName(specDetails.t1.displayName)
                            spec.setAvatar(specDetails.t2)
                        }
                    }
                    .flatMap { webhook ->
                        webhook.execute { spec ->
                            spec.setContent(targetMessage.content)
                            targetMessage.attachments.forEach { spec.addFile(it.filename, BotUtils.getAttachment(it.url)) }
                            targetMessage.embeds.forEach { embed ->
                                spec.addEmbed { embedSpec ->
                                    embed.author.ifPresent { author -> embedSpec.setAuthor(author.name.get(), author.url.orElse(null), author.iconUrl.orElse(null)) }
                                    embed.color.ifPresent { embedSpec.setColor(it) }
                                    embed.description.ifPresent { embedSpec.setDescription(it) }
                                    embed.fields.forEach { field -> embedSpec.addField(field.name, field.value, field.isInline) }
                                    embed.footer.ifPresent { footer -> embedSpec.setFooter(footer.text, footer.iconUrl.orElse(null)) }
                                    embed.image.map { it.url }.ifPresent { embedSpec.setUrl(it) }
                                    embed.thumbnail.map { it.url }.ifPresent { embedSpec.setThumbnail(it) }
                                    embed.timestamp.ifPresent { embedSpec.setTimestamp(it) }
                                    embed.title.ifPresent { embedSpec.setTitle(it) }
                                    embed.url.ifPresent { embedSpec.setUrl(it) }
                                }
                            }
                        }.then(webhook.delete())
                    }
                    .doOnSuccess {
                        targetMessage.delete("Moved message").subscribe()
                        commandManager.sendResponse(message.channel, "Message moved", "The message was moved successfully.").subscribe()
                    }.subscribe()
        }
    }

}
